### PR TITLE
Add a couple more blocks to deco phase

### PR DESCRIPTION
**Pull request by Mekle001**
# Changes proposed in this pull request:
- Add Banners to deco phase of build
- Add Wall signs to deco phase of build
-

Review please

### CODE REVIEW

**Comment by Raycoms** [comment]
:shipit: 

[Approved with PullApprove](https://pullapprove.com/ldtteam/minecolonies/pull-request/5975/?utm_source=github-pr&utm_medium=comment-badge&utm_campaign=ldtteam/minecolonies)

File: src/main/java/com/minecolonies/coremod/entity/ai/basic/AbstractEntityAIStructure.java [hunk 1]
```diff
@@ -238,6 +238,14 @@ public boolean walkToConstructionSite(final BlockPos currentBlock)
         return worker.isWorkerAtSiteWithMove(workFrom, STANDARD_WORKING_RANGE) || MathUtils.twoDimDistance(worker.getPosition(), workFrom) < MIN_WORKING_RANGE;
     }
 
+    /**
+     * Checks for blocks that need to be treated as deco
+     */
+    protected boolean isDecoItem(Block block)
+    {
+        return block instanceof CoralBlock || block instanceof LanternBlock || block instanceof BannerBlock || block instanceof WallSignBlock;
```
Comment (marchermans) [inline src/main/java/com/minecolonies/coremod/entity/ai/basic/AbstractEntityAIStructure.java]: Yeah no, this is now officially way to much OR-InstanceOf combos.
Add a config entry todo this, or add a simple list based registry in the api, regardless this is not a great maintainable approach anymore.

Comment (marchermans) [reply src/main/java/com/minecolonies/coremod/entity/ai/basic/AbstractEntityAIStructure.java]: Or add a Tag?
That would probably be the best approach.

**Review by marchermans — CHANGES_REQUESTED** [review]

File: src/main/java/com/minecolonies/coremod/entity/ai/basic/AbstractEntityAIStructure.java (hunk 1 above)
Comment (Raycoms) [reply src/main/java/com/minecolonies/coremod/entity/ai/basic/AbstractEntityAIStructure.java]: Yeah, a tag is probably the best thing.

File: src/main/java/com/minecolonies/coremod/entity/ai/basic/AbstractEntityAIStructure.java [hunk 2]
```diff
@@ -147,6 +160,8 @@ protected AbstractEntityAIStructure(@NotNull final J job)
            */
           new AITarget(COMPLETE_BUILD, this::completeBuild, STANDARD_DELAY)
         );
+
+        decoItems = BlockTags.getCollection().getOrCreate(new ResourceLocation(MOD_ID, DECOBLOCK_TAG));
```
Comment (marchermans) [inline src/main/java/com/minecolonies/coremod/entity/ai/basic/AbstractEntityAIStructure.java]: This is not the way to do tags.

**Review by marchermans — CHANGES_REQUESTED** [review]

File: src/main/java/com/minecolonies/coremod/entity/ai/basic/AbstractEntityAIStructure.java (hunk 2 above)
Comment (Mekle001) [reply src/main/java/com/minecolonies/coremod/entity/ai/basic/AbstractEntityAIStructure.java]: Ok, I've updated this to static, to use a better pattern for the caching of this. Given that this is the way all of our tags access is in 1.15, I think it should stay consistent here and be updated to be 'right' in 1.16 using the TagUtils class.

File: src/main/java/com/minecolonies/coremod/entity/ai/basic/AbstractEntityAIStructure.java [hunk 3]
```diff
@@ -80,6 +83,16 @@
...
 
+    /**
+     * Tagname for decoration blocks
+     */
+    private static String DECOBLOCK_TAG = "decoblocks";
+
+    /**
+     * Tag of all the blocks that need to be in the deco phase of the builder
+     */
+    protected static Tag<Block> decoItems;
```
Comment (marchermans) [inline src/main/java/com/minecolonies/coremod/entity/ai/basic/AbstractEntityAIStructure.java]: No. As discussed moved the actual tag instance to a dedicated class in the API which will group all of these.

**Review by marchermans — CHANGES_REQUESTED** [review]

File: src/main/java/com/minecolonies/coremod/entity/ai/basic/AbstractEntityAIStructure.java (hunk 3 above)
Comment (Raycoms) [reply src/main/java/com/minecolonies/coremod/entity/ai/basic/AbstractEntityAIStructure.java]: Yeah, but that's the way we do this in 1.15 everywhere (also in several PR that were approved from you in the past). In the 1.16 port of this PR we can move this there

Comment (Mekle001) [reply src/main/java/com/minecolonies/coremod/entity/ai/basic/AbstractEntityAIStructure.java]: I think we all are now on the same page around what you're asking for Orion, with something like a ModTags module. 

I've been thinking about it overnight, and I realized we CAN NOT simply cache these returns, as they could be redefined at any time by a /reload command reloading a new datapack. 

This means that the proposed ModTags module isn't really following the rest of our Mod* pattern.

The "best" way I can currently see is a module with getters that lazy init the instances, and a dummy reload listener that clears the caches, with a complete sweep of the codebase to update to use it. Is that really a change we want to take in 1.15, or should it be done in 1.16 as a cleanup item?

Comment (Raycoms) [reply src/main/java/com/minecolonies/coremod/entity/ai/basic/AbstractEntityAIStructure.java]: There is an event to do that. So we can add an event listener to that ModTags class that handles that.

File: src/api/java/com/minecolonies/api/items/ModTags.java [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+import net.minecraft.tags.Tag;
+
+public class ModTags
+{
+    public static Tag<Block>    decorationItems;
+    public static Tag<Item>     concretePowder;
+    public static Tag<Block>    concreteBlock;
+    public static Tag<Block>    pathingBlocks;
+
+    public static Map<String, Tag<Item>> crafterProduct;
```
Comment (marchermans) [inline src/api/java/com/minecolonies/api/items/ModTags.java]: Initialize this straight up as a final hashmap.

File: src/main/java/com/minecolonies/apiimp/initializer/ModTagsInitializer.java [hunk 5]
```diff
@@ -0,0 +1,137 @@
...
+    private static final ResourceLocation DECORATION_ITEMS = new ResourceLocation(MOD_ID, "decoblocks");
+    private static final ResourceLocation CONCRETE_POWDER = new ResourceLocation(MOD_ID, "concrete_powder");
+    private static final ResourceLocation CONCRETE_BLOCK  = new ResourceLocation(MOD_ID, "concrete");
+    private static final ResourceLocation PATHING_BLOCKS = new ResourceLocation(MOD_ID, "pathblocks");
+
+
+    @SubscribeEvent
+    public static void tagsUpdated(TagsUpdatedEvent event)
+    {
+        ModTagsInitializer.init();
```
Comment (marchermans) [inline src/main/java/com/minecolonies/apiimp/initializer/ModTagsInitializer.java]: Not needed.
We use named tags as such we do not need this.

**Review by marchermans — CHANGES_REQUESTED** [review]
Looking much better.
Some minor nitpickings, since this still will crash on 1.16 due to the named tag being pulled multiple times (aka during TagReload) That even is not needed anyway, since we use named tags which are only a wrapper around an inner tag object which gets automatically switched out.

File: src/main/java/com/minecolonies/apiimp/initializer/ModTagsInitializer.java (hunk 5 above)
Comment (Raycoms) [reply src/main/java/com/minecolonies/apiimp/initializer/ModTagsInitializer.java]: Apparently a static initializer is enough here.

**Review by marchermans — APPROVED** [review]

**Pull request merged**